service/storage_service.cc

```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
                 }
             } else if (req.cmd == node_ops_cmd::removenode_abort) {
                 ss.node_ops_abort(ops_uuid);
+            } else if (req.cmd == node_ops_cmd::replace_prepare) {
+                // Mark the replacing node as replacing
+                if (req.replace_nodes.size() > 1) {
+                    auto msg = format("replace[{}]: Could not replace more than one node at a time: replace_nodes={}", req.ops_uuid, req.replace_nodes);
+                    slogger.warn("{}", msg);
+                    throw std::runtime_error(msg);
+                }
+                ss.mutate_token_metadata([coordinator, &req, &ss] (mutable_token_metadata_ptr tmptr) mutable {
+                    for (auto& x: req.replace_nodes) {
+                        auto existing_node = x.first;
+                        auto replacing_node = x.second;
+                        slogger.info("replace[{}]: Added replacing_node={} to replace existing_node={}, coordinator={}", req.ops_uuid, replacing_node, existing_node, coordinator);
+                        tmptr->add_replacing_endpoint(existing_node, replacing_node);
```

Comment:
This is the key point in the solution. After adding replacing node, the replacing node will be excluded from the normal read and write, i.e., other nodes will not send either read or write requests to replacing node.

Reply:
What if that node restarts after we executed this step? Will it revert to the previous state?

Reply:
The participant node do not remember the state, so no. But if the node restarts, the replacing operation will fail anyway. The cluster will revert. So the node does not need to revert to the state with add_replacing_endpoint executed.

Reply:
What guarantees that the operation will fail?

Before it fails, is it possible that the node which restarted will use the wrong (old) topology, and will try to read from the new node as if it was the old node being replaced?

Reply:
The rpc connections to the replacing node will be killed when the replacing node detects a node is restarted, so the streaming will fail and in turn the replace operation will fail.

If the streaming process is already finished before the node restarts, it is not a big problem because the replacing node already have all the data. 

In addition, since the restarted node does not know the ops_uuid, the restarted node will fail any cmd sent by the replacing node, in turn fails the node operation.

Reply:
What if the node restarts before we start streaming?

What about the second concern about consistency?

Reply:
```
In addition, since the restarted node does not know the ops_uuid, the restarted node will fail any cmd sent by the replacing node, in turn fails the node operation.
```

See v2.

Reply:
The replace operation will eventually fail, but we still will run into consistency problems until it does, because the replacing node will respond to requests as if it was the old node, sent by the node which restarted and is using the old topology.

The problem here is that the state transitions are not monotonic, restarts revert them, but the solution depends on them being monotonic. If status was propagated via gossip, they would be monotonic.

Reply:
The problem with gossip propagation can be even worse, e.g., the other node can completely miss the gossip propagation.

So the problem here is

- replacing node uses the same ip address of the node to be replaced
- in the middle of the replace operation one of the node restarted and the restarted node will mark the replacing node as up

Let's break it down:

// Step 1: Decide who needs to sync data for replace operation
// Step 2: Prepare to sync data
// Step 3: Start heartbeat updater
// Step 4: Allow nodes to mark the replacing node as alive
_gossiper.advertise_myself().get();
// Step 5: Wait for nodes to finish marking the replacing node as live
// Step 6: Update pending ranges on nodes
// Step 7: Sync data for replace
// Step 8: Finish

If the restart happens before Step 4, it is not a problem. The replacing node will not respond to mark alive at all. 

If the restart happens after Step 4, we already have the heartbeat updater, The restarted node will fail the heart beat update. We can fail the replace operation.

Reply:
I don't think you addressed my concern about potential consistency problems. Even if replace operation eventually fails, some requests may use the wrong topology in the mean time.

Reply:
See v3. The restarted node will not mark the replacing node as alive.